File: SimpleEngine/ActorFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;
using System.IO;

namespace SimpleEngine
{
    public static class ActorFile
    {
        public static Actor Load(FileSystem fs, string p)
        {
            XmlElement xactor = Xml.Open(Xml.FromStream(fs.open(p)), "actor");
            XmlElement xmesh = xactor["mesh"];

            List<AnimationInformation> animinfo = ParseAnimationInfo(xmesh);
            List<string> bonesToIgnore = ParseIgnoreBone(xmesh);

            Dictionary<string, string> texmap = ParseSetTexture(xmesh);
            Dictionary<string, string> overrides = ParseOverrides(xmesh);

            fs.setOverrides(overrides);

            string meshpath = Xml.GetAttributeString(xmesh, "file");
            float scale =  Xml.GetAttribute<float>(xmesh, "scale", math1.ParseFloat, 1.0f);

            MeshDef def;
            Animation animation;

            string ext = Path.GetExtension(meshpath);
            if (ext == ".txt")
            {
                SimpleEngine.load.MilkshapeAscii.Load(fs, meshpath, out def, out animation, 1);
            }
            else if (ext == ".ms3d")
            {
                SimpleEngine.load.MilkshapeBinary.Load(fs, meshpath, out def, out animation);
            }
            else if( animinfo.Count==0 )
            {
                animation = null;
                def = MeshFile.Load(fs, meshpath);
            }
            else throw new Exception("Unhandled format " + ext + " for " + meshpath);

            foreach (string ignoreThisBone in bonesToIgnore)
            {
                for (int i = 0; i < def.bones.Count; ++i)
                {
                    if (ignoreThisBone == def.bones[i].name)
                    {
                        def.bones.RemoveAt(i);
                        animation.bones.RemoveAt(i);
                    }
                }
            }

            foreach (MeshDef.MaterialDef mat in def.Materials)
            {
                string matname = mat.name.ToLower();
                if (texmap.ContainsKey(matname))
                {
                    mat.texture = texmap[matname];
                    texmap.Remove(matname);
                }
            }

            if (texmap.Count != 0) throw new Exception("Some materials was not mapped");

            def.scale(scale);
            if( animation != null ) animation.scale(scale);

            def.translateFiles(overrides);

            Actor actor = new Actor(def.mapBones());

            if (animation != null)
            {
                foreach (AnimationInformation ai in animinfo)
                {
                    Animation an = animation.subanim(ai);
                    actor.add(ai.name, an);
                }
            }

            fs.clearOverrides(overrides);
            return actor;
        }

        private static Dictionary<string, string> ParseOverrides(XmlElement root)
        {
            Dictionary<string, string> result = new Dictionary<string, string>();
            foreach (XmlElement e in Xml.ElementsNamed(root, "overrride"))
            {
                string f = Xml.GetAttributeString(e, "from").ToLower();
                string t = Xml.GetAttributeString(e, "to").ToLower();
                result.Add(f, t);
            }
            return result;
        }

        private static List<AnimationInformation> ParseAnimationInfo(XmlElement root)
        {
            List<AnimationInformation> animinfo = new List<AnimationInformation>();
            foreach (XmlElement e in Xml.ElementsNamed(root, "animation"))
            {
                string ss = Xml.GetTextOfSubElement(e, "start");
                string se = Xml.GetTextOfSubElement(e, "end");
                string name = Xml.GetTextOfSubElement(e, "name");
                int start = int.Parse(ss);
                int end = int.Parse(se);
                animinfo.Add(new AnimationInformation(start, end, name));
            }
            return animinfo;
        }

        private static List<string> ParseIgnoreBone(XmlElement root)
        {
            List<string> bonestoignore = new List<string>();
            foreach (XmlElement e in Xml.ElementsNamed(root, "ignorebone"))
            {
                string ss = Xml.GetAttributeString(e, "name");
                bonestoignore.Add(ss);
            }
            return bonestoignore;
        }

        private static Dictionary<string, string> ParseSetTexture(XmlElement root)
        {
            Dictionary<string, string> texmap = new Dictionary<string, string>();
            foreach (XmlElement e in Xml.ElementsNamed(root, "settex"))
            {
                string material = Xml.GetAttributeString(e, "material");
                string texture = Xml.GetAttributeString(e, "texture");
                texmap.Add(material, texture);
            }
            return texmap;
        }
    }
}

File: src/engine/actorfile.cc
#include <iostream>
#include <map>
#include <set>
#include <stdexcept>
#include <vector>

#include "engine/actor.h"
#include "engine/animation.h"
#include "engine/cpp.h"
#include "engine/filesystem.h"
#include "engine/fileutil.h"
#include "engine/load/milkshape.ascii.h"
#include "engine/load/milkshape.binary.h"
#include "engine/math.h"
#include "engine/meshfile.h"
#include "engine/strings.h"
#include "engine/xml.h"
#include "fmt/core.h"

namespace simple
{
    namespace actor_file
    {
        std::map<std::string, std::string> ParseOverrides(std::shared_ptr<xml::Element> root)
        {
            std::map<std::string, std::string> result;
            for (std::shared_ptr<xml::Element> e : xml::ElementsNamed(root, "overrride"))
            {
                const auto f = ToLower(xml::GetAttributeString(e, "from"));
                const auto t = ToLower(xml::GetAttributeString(e, "to"));
                result.emplace(f, t);
            }
            return result;
        }

        std::vector<AnimationInformation> ParseAnimationInfo(std::shared_ptr<xml::Element> root)
        {
            std::vector<AnimationInformation> animinfo;
            for (std::shared_ptr<xml::Element> e : xml::ElementsNamed(root, "animation"))
            {
                const auto ss = xml::GetTextOfSubElement(e, "start");
                const auto se = xml::GetTextOfSubElement(e, "end");
                const auto name = xml::GetTextOfSubElement(e, "name");
                int start = std::stoi(ss);
                int end = std::stoi(se);
                animinfo.push_back(AnimationInformation(start, end, name));
            }
            return animinfo;
        }

        std::set<std::string> ParseIgnoreBone(std::shared_ptr<xml::Element> root)
        {
            std::set<std::string> bonestoignore;
            for (std::shared_ptr<xml::Element> e : xml::ElementsNamed(root, "ignorebone"))
            {
                const auto ss = xml::GetAttributeString(e, "name");
                bonestoignore.emplace(ss);
            }
            return bonestoignore;
        }

        std::map<std::string, std::string> ParseSetTexture(std::shared_ptr<xml::Element> root)
        {
            std::map<std::string, std::string> texmap;
            for (std::shared_ptr<xml::Element> e : xml::ElementsNamed(root, "settex"))
            {
                const auto material = xml::GetAttributeString(e, "material");
                const auto texture = xml::GetAttributeString(e, "texture");
                texmap.emplace(material, texture);
            }
            return texmap;
        }

        std::shared_ptr<Actor> Load(FileSystem* fs, const std::string& p)
        {
            const auto full_path = fs->Open(p);
            std::shared_ptr<xml::Element> xactor = xml::Open(full_path, "actor");
            std::shared_ptr<xml::Element> xmesh = xactor->GetChild("mesh");

            const auto animinfo = ParseAnimationInfo(xmesh);
            const auto bonesToIgnore = ParseIgnoreBone(xmesh);
            auto texmap = ParseSetTexture(xmesh);
            const auto overrides = ParseOverrides(xmesh);

            fs->SetOverrides(overrides);

            const auto meshpath = xml::GetAttributeString(xmesh, "file");
            float scale = xml::GetAttribute<float>(xmesh, "scale", math::ParseFloat, 1.0f);

            std::shared_ptr<MeshDef> def;
            std::shared_ptr<Animation> animation;

            std::string ext = file_util::GetExtension(meshpath);
            if (ext == ".txt")
            {
                simple::load::milkshape_ascii::Load(fs, meshpath, &def, &animation, 1);
            }
            else if (ext == ".ms3d")
            {
                simple::load::milkshape_binary::Load(fs, meshpath, &def, &animation);
            }
            else if (animinfo.empty())
            {
                animation = nullptr;
                def = mesh_file::Load(fs, meshpath);
            }
            else
            {
                throw std::runtime_error(fmt::format("Unhandled format {} for {}", ext, meshpath));
            }

            for (std::string ignoreThisBone : bonesToIgnore)
            {
                for (int i = 0; i < def->bones.size(); ++i)
                {
                    if (ignoreThisBone == def->bones[i]->name)
                    {
                        def->bones.erase(def->bones.begin() + i);
                        animation->bones.erase(animation->bones.begin() + i);
                        std::cout << "bug? reduce i by 1 since it was just removed?\n";
                    }
                }
            }

            for (auto mat : def->GetMaterials())
            {
                std::string matname = ToLower(mat->name);
                if (auto found = texmap.find(matname); found != texmap.end())
                {
                    mat->texture = found->second;
                    texmap.erase(found);
                }
            }

            if (texmap.empty() == false)
                throw std::runtime_error("Some materials was not mapped");

            def->ScaleMeshAndBones(scale);
            if (animation != nullptr)
                animation->Scale(scale);

            def->TranslateTexturePaths(overrides);

            def->MapBones();
            std::shared_ptr<Actor> actor = std::make_shared<Actor>(def);

            if (animation != nullptr)
            {
                for (AnimationInformation ai : animinfo)
                {
                    auto an = animation->GetSubAnimation(ai);
                    actor->AddAnimation(ai.name, std::make_shared<Animation>(an));
                }
            }

            fs->ClearOverrides(overrides);
            return actor;
        }
    }
}
